notes/learning_java.md

```diff
@@ -0,0 +1,462 @@
+
+`Unbounded wildcard types`: If you want to use a generic type but you don’t know or care what the actual type parameter is, you can use a question mark instead. For example, the unbounded wildcard type for the generic type Set<E> is Set<?> (read “set of some type”). It is the most general parameterized Set type, capable of holding any set.
+
+## Generics: Terms and Examples
+
+| Term                    | Example                          |
+|-------------------------|----------------------------------|
+| Parameterized type      | List<String>                     |
+| Actual type parameter   | String                           |
+| Generic type            | List<E>                          |
+| Formal type parameter   | E                                |
+| Unbounded wildcard type | List<?>                          |
+| Raw type                | List                             |
+| Bounded type parameter  | <E extends Number>               |
+| Recursive type bound    | <T extends Comparable<T>>        |
+| Bounded wildcard type   | List<? extends Number>           |
+| Generic method          | static <E> List<E> asList(E[] a) |
+| Type token              | String.class                     |
+
+
+# Collections
+
+Java provides Collection Framework which defines several classes and interfaces to represent a group of objects as a single unit.
+
+The Collection interface (java.util.Collection) and Map interface (java.util.Map) are two main root interfaces of Java collection classes.
+
+`Hierarchy of Collection Framework` (core interfaces only)
+
+```
+             Collection                Map
+         /     /    \      \            |
+        /      /      \     \           |
+     Set    List    Queue  Dequeue   SortedMap
+     /
+    /
+ SortedSet
+```
+---
+
+```
+Collection : Root interface with basic methods like add(), remove(),
+             contains(), isEmpty(), addAll(), ... etc.
+
+Set : Doesn't allow duplicates. Example implementations of Set
+      interface are HashSet (Hashing based) and TreeSet (balanced
+      BST based). Note that TreeSet implements SortedSet.
+
+List : Can contain duplicates and elements are ordered. Example
+       implementations are LinkedList (linked list based) and
+       ArrayList (dynamic array based)
+
+Queue : Typically order elements in FIFO order except exceptions
+        like PriorityQueue.
+
+Deque : Elements can be inserted and removed at both ends. Allows
+        both LIFO and FIFO.
+
+Map : Contains Key value pairs. Doesn't allow duplicates.  Example
+      implementation are HashMap and TreeMap.
+      TreeMap implements SortedMap.
+
+The difference between Set and Map interface is, in Set we have only
```

Comment:
If you remember what a Map is then you can think of a Set as being a Map without the value bit. Or, in even more casual terms, 'the left hand side of a map'